### ISSUE TITLE
NetworkPeerConnectionParameters.CreateVersion: System.NullReferenceException

### ISSUE THREAD

**Issue by sondreb** [stratisproject/StratisBitcoinFullNode#2447]
The NetworkPeer constructor will fail when it calls this line:

```
this.MyVersion = this.ConnectionParameters.CreateVersion(this.selfEndpointTracker.MyExternalAddress, this.PeerEndPoint, network, this.dateTimeProvider.GetTimeOffset());
```

this.selfEndpointTracker.MyExternalAddress is NULL, which makes the "new VersionPayload()" inside that CreateVersion method, fail. Perhaps the CreateVersion should check for nulls?

Looking at the SelfEndpointTracker, and debugging, it appears that MyExternalAddress is read before the UpdateAndAssignMyExternalAddress is executed. So the value is never set, before it is attempted to be access by NetworkPeer.

I do think it will recover from this, cause this appears a few lines down in the log:

```
Node listening on: 0.0.0.0:24333
```
I don't know for sure, but from the log it appears to not be a very serious problem, but should still avoid exceptions if possible.

**Comment by sondreb** [comment]
I have experienced this issue more and did more debugging, I found that the problem is a race-condition that happens.

If you debug, it is fairly easy to replicate, but at runtime it happens at random since it depends on how many connections you have, and how fast they connect, etc.

If we start at the problem-end (where an exception happens in the constructor), and roll backwards, it is as followed:

```
var version = new VersionPayload() is called by public VersionPayload CreateVersion.
```

VersionPayload is a NBitcoin type, and it will throw exception is AddressFrom or AddressReceiver is set to null, this is because it attempts to parse this into an IPEndpoint.

The problem is the externalAddressEndPoint parameter. This can during startup, sometimes be null, due to race condition.

Rollup backwards, the CreateVersion method is called from the constructor of the NetworkPeer class.
```
this.MyVersion = this.ConnectionParameters.CreateVersion
```
Since there are no try/catch in the constructor, this exception flows backward (upwards) to: CreateConnectedNetworkPeerAsync

![image](https://user-images.githubusercontent.com/309938/49592186-bf8ad400-f970-11e8-9374-efce5bc31eb7.png)

Which does not wrap the object construction into a try/catch. This is an async method, which is called by CreateConnectedNetworkPeerAsync (right above).

This is yet another async method, and it too simply returns the Task returned from overloaded method.

Which is called within the await peersToDiscover.ForEachAsync in PeerDiscoveryLoop.cs:

networkPeer = await this.networkPeerFactory.CreateConnectedNetworkPeerAsync(endPoint, clonedParameters).ConfigureAwait(false);

This does encapsulate the call in a try/catch/finally, but swallows the exception with no logging, and then dispose the networkPeer, if it exists (which it does not as it failed creation.

It does though log null reference exception when running trace logging.

Here is the Initialize method on ConnectionManager. The problem is that the .MyVersion assignment is happening BEFORE the last breakpoint shown in this screenshot. And the MyExternalAddress is not set until this method is called.

![image](https://user-images.githubusercontent.com/309938/49592358-30ca8700-f971-11e8-8843-731363371029.png)

One possible solution could be to set the external IP already when this.AddExternalIpToSelfEndpoints(); is called. And perhaps update it later on below. This should help the problem.

**Comment by MithrilMan** [comment]
I think this deserves some attention because may be the cause of weird connection statuses

**Comment by noescape00** [comment]
@dangershony @sondreb is this issue still valid?

**Comment by dangershony** [comment]
Yes I beleive it is still an issue, mostly on first startup.

**Comment by maciejzaleski** [comment]
Resolved by #3370
